File: dentis-frontend/src/app/core/services/budget.service.ts
import { Injectable } from '@angular/core';
import { HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { ApiService, PageResponse } from './api.service';
import {
  Budget, CreateBudgetRequest, CreatePaymentRequest, Payment
} from '../models/budget.model';

@Injectable({ providedIn: 'root' })
export class BudgetService {
  constructor(private api: ApiService) {}

  create(request: CreateBudgetRequest): Observable<Budget> {
    return this.api.post<Budget>('/budgets', request);
  }

  findById(id: string): Observable<Budget> {
    return this.api.get<Budget>(`/budgets/${id}`);
  }

  findByPatient(patientId: string, page = 0, size = 20): Observable<PageResponse<Budget>> {
    const params = new HttpParams().set('page', page).set('size', size);
    return this.api.get<PageResponse<Budget>>(`/budgets/patient/${patientId}`, params);
  }

  approve(id: string): Observable<Budget> {
    return this.api.post<Budget>(`/budgets/${id}/approve`, {});
  }

  reject(id: string): Observable<Budget> {
    return this.api.post<Budget>(`/budgets/${id}/reject`, {});
  }

  markItemPerformed(budgetId: string, itemId: string): Observable<void> {
    return this.api.patch<void>(`/budgets/${budgetId}/items/${itemId}/performed`);
  }

  registerPayment(request: CreatePaymentRequest): Observable<Payment> {
    return this.api.post<Payment>('/payments', request);
  }

  getPayments(budgetId: string): Observable<Payment[]> {
    return this.api.get<Payment[]>(`/payments/budget/${budgetId}`);
  }
}

File: dentis-application/src/main/java/com/dentis/application/service/BudgetService.java
package com.dentis.application.service;

import com.dentis.application.dto.request.CreateBudgetRequest;
import com.dentis.application.dto.response.BudgetResponse;
import com.dentis.common.response.PageResponse;
import com.dentis.domain.enums.BudgetStatus;
import org.springframework.data.domain.Pageable;

public interface BudgetService {

    BudgetResponse create(CreateBudgetRequest request);

    BudgetResponse findById(String id);

    PageResponse<BudgetResponse> findByPatient(String patientId, Pageable pageable);

    BudgetResponse approve(String id);

    BudgetResponse reject(String id);

    void markItemPerformed(String budgetId, String itemId);
}
